### DISCUSSION TITLE
Open a GitHub repo (including org and private repos, as well as pagination)

### DISCUSSION THREAD

**Discussion by mandiwise** [johnlindquist/kit#252] [Share]
This script allows you to authenticate with a GitHub personal access token and open repos from your user account or any organizations you belong to. It also displays private repos and handles pagination if the repo count exceeds 100 (just scroll to the bottom of the list and you'll see a "Load more..." option, when applicable).

![Opening an org repo](https://user-images.githubusercontent.com/1518780/117495102-70c33900-af32-11eb-9e04-be54ca374032.gif)

```js
// Menu: Open a GitHub Repo
// Description: Launch a GitHub repo in your browser
// Author: Mandi Wise
// Twitter: @mandiwise

// Learn how to create a personal access token for GitHub here:
// https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token

let { Octokit } = await npm("octokit");

const GH_PERSONAL_ACCESS_TOKEN = await env("GH_PERSONAL_ACCESS_TOKEN");
const octokit = new Octokit({ auth: GH_PERSONAL_ACCESS_TOKEN });

const ORGS_PER_PAGE = 20;
const REPOS_PER_PAGE = 100;

// GraphQL operations

const GetAccounts = `query GetAccounts($first: Int) { 
  viewer { 
    organizations(first: $first) {
      edges {
        node {
          login
          name
          url
        }
      }
    }
    login
    name
    url
  }
}`;

const ReposPage = `fragment ReposPage on RepositoryConnection {
  edges {
    node {
      name
      description
      url
    }
  }
  pageInfo {
    endCursor
    hasNextPage
  }
}`;

const GetOrgRepos = `query GetOrgRepos($first: Int, $after: String, $login: String!) {
  viewer { 
    organization(login: $login) {
      repositories(
        first: $first
        after: $after
        orderBy: { field: UPDATED_AT, direction: DESC }
      ) {
        ...ReposPage
      }
    }
  }
}
${ReposPage}
`;

const GetUserRepos = `query GetUserRepos($first: Int, $after: String) {
  viewer { 
    repositories(
      first: $first
      after: $after
      orderBy: {field: UPDATED_AT, direction: DESC}
      affiliations: OWNER
    ) {
      ...ReposPage
    }
  }
}
${ReposPage}
`;

// Get user and their organizations in a list

let dots = 0;
const accountsPlaceholderIntervalId = setInterval(() => {
  setPlaceholder(`Loading GitHub accounts`.padEnd(++dots, "."));
}, 100);

const { viewer } = await octokit.graphql(GetAccounts, { first: ORGS_PER_PAGE });

if (!viewer) {
  exit(1);
}

const { login, name, url, organizations } = viewer;
const accounts = [
  { name, value: login, description: url, type: "user" },
  ...organizations.edges.map(({ node: { login, name, url } }) => ({
    name,
    value: login,
    description: url,
    type: "org"
  }))
].sort((a, b) => (a.name > b.name ? 1 : -1));

clearInterval(accountsPlaceholderIntervalId);
dots = 0;

const accountChoice = await arg("Which account?", accounts);
const { type: accountType } = accounts.find(
  account => accountChoice === account.value
);

// Get repo list for the user or organization

let repositoriesAndLoadMore = [];

async function fetchRepositories(variables) {
  const reposPlaceholderIntervalId = setInterval(() => {
    setPlaceholder(`Loading repositories`.padEnd(++dots, "."));
  }, 100);

  let edges, endCursor, hasNextPage;
  const oldLoadMore = repositoriesAndLoadMore.find(({ value }) =>
    value.startsWith("load-more-after-")
  );

  if (oldLoadMore) {
    repositoriesAndLoadMore.pop();
  }

  if (accountType === "org") {
    ({
      viewer: {
        organization: {
          repositories: {
            edges,
            pageInfo: { endCursor, hasNextPage }
          }
        }
      }
    } = await octokit.graphql(GetOrgRepos, {
      login: accountChoice,
      ...variables
    }));
  } else {
    ({
      viewer: {
        repositories: {
          edges,
          pageInfo: { endCursor, hasNextPage }
        }
      }
    } = await octokit.graphql(GetUserRepos, variables));
  }

  repositoriesAndLoadMore = [
    ...repositoriesAndLoadMore,
    ...edges.map(({ node: { description, name, url } }) => ({
      name,
      description,
      value: url
    }))
  ];

  if (hasNextPage) {
    repositoriesAndLoadMore.push({
      name: "Load more...",
      value: `load-more-after-${endCursor}`
    });
  }

  clearInterval(reposPlaceholderIntervalId);
  dots = 0;

  if (!repositoriesAndLoadMore.length) {
    exit(1);
  }

  let repoChoice = await arg("Which project?", repositoriesAndLoadMore);

  if (repoChoice.startsWith("load-more-after-")) {
    await fetchRepositories({
      first: REPOS_PER_PAGE,
      after: repoChoice.split("-").pop()
    });
  } else {
    exec(`open ${repoChoice}`);
  }
}

await fetchRepositories({ first: REPOS_PER_PAGE });
```

**Comment by gr2m** [comment]
Hey @mandiwise, very cool script! Paginating GraphQL queries is tricky, but we do work on a plugin for that, in case you'd like to add automated pagination in future:
https://github.com/octokit/plugin-paginate-graphql.js/pull/1

You might also be interested in https://github.com/gr2m/scriptkit-octokit/, which is the `octokit` module you already use, plus the [OAuth Device authentication strategy](https://github.com/octokit/auth-oauth-device.js), so you don't need to prompt users to create tokens manually.

If you have any Octokit related questions, please let me know :)

**Reply by mandiwise** [reply]
Thanks for the tip, this looks like it will definitely simplify things!

**Comment by gr2m** [comment]
I tried running your script but it opens my editor, see the video. I'm using v1.3.0-beta.12 /cc @johnlindquist 


https://user-images.githubusercontent.com/39992/117503215-7f5f1f80-af35-11eb-9e87-9f8b55f4cb8c.mp4

**Reply by johnlindquist** [reply]
Investigating...

(When an error occurs, Script Kit attempts to open the editor and jump to the line number based on the `error.stack`. Obviously it's far from perfect 😅)

**Reply by mandiwise** [reply]
@gr2m That's strange, I noticed similar behavior when I initially had an authentication error with the GitHub GraphQL API.

I was able to debug the issue with the logs in `~/.kenv/logs`. Let me know if you find anything interesting there.

**Reply by johnlindquist** [reply]
@gr2m I had to give the token the "admin:org" scope, then it worked.

**Reply by johnlindquist** [reply]
@mandiwise btw, glad I installed this. Super useful!

**Reply by gr2m** [reply]
It works with just `read:org`. 

@mandiwise here is the code change I'd reccomend

```diff
- let { Octokit } = await npm("octokit");
- 
- const GH_PERSONAL_ACCESS_TOKEN = await env("GH_PERSONAL_ACCESS_TOKEN");
- const octokit = new Octokit({ auth: GH_PERSONAL_ACCESS_TOKEN });
+ let { Octokit } = await npm("scriptkit-octokit");
+ const octokit = new Octokit({ auth: { scopes: ["read:org"] } });
```

@johnlindquist it should fail more gracefully, not sure if it's Mandi's script, octokit, or kit. I'll have a look

**Reply by johnlindquist** [reply]
@gr2m Yeah, it's Script Kit failing that way. It parses the error stack to open the file. It works well with errors in scripts, but fails like your video when the error is Thrown in an npm package. I'll get it sorted.

**Reply by gr2m** [reply]
The first GraphQL query fails with the following error

> Your token has not been granted the required scopes to execute this query. The 'login' field requires one of the following scopes: ['read:org'], but your token has only been granted the: [''] scopes

You can wrap your whole code in a try/error, and set the `accountsPlaceholderIntervalId` outside of it, so that you can clear the interval in case of both success and error.

But I wonder @johnlindquist if it would make sense to wrap the whole script in try/error, and log the error in case the script fails, with some useful UI and hints on how to debug it?

**Reply by johnlindquist** [reply]
@gr2m yeah, I have a lot of work to do with improving the error-handling experience. Just give me some time. I have a lot of tasks on my plate :)
